keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+
+Moreover, people building declarative application description with several
+Kubernetes resources cannot easily centralize clean up using owner references,
+meaning, to create a bunch of objects and deleting one of them to make the other
+objects to go away with it. 
+
+One of the reasons is that owner resource [UID][] is required when specifying
+owner references, but UID is unknown until after a resource is created. Even if
+Kubernetes removes this limitation and allow named based owner references,
+there’s still a resource creation ordering problem: any child resource created
+before its parent resource is prone to be cleaned up by the garbage collector
+unexpectedly.
+
+What's more, a child resource may depend on many other parent resources. It's
+likely for the list of owners of a child resource to be excessively long, and
+hit Kubernetes resource size limit. The resource then cannot be updated anymore
+and becomes useless. 
+
+[UID]: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids
+
+## Proposal
+
+The design has three main parts: 
+
+1. **Object-lifecycle-respecting TTL mechanism.** Provide observability before
+   deletion, and address the resource creation ordering problems. See [detailed
+   design of TTL][].
+1. **Garbage collect resources that finish execution.** Delete a resource object
+   when it finishes execution (either terminates or completes). See [detailed
+   design of GC after use (part 1)][].
+1. **Garbage collect resources that are not used by other resources.** Extract
+   usage references of a resource object and delete the object when it is no
+   longer referenced. Garbage collector will count usage references and address
+   the problem of unbounded list of owner references. See [detailed design of GC
+   after use (part 2)][].
+
+This will be coherent with existing owner-reference-based garbage collection
+mechanism. This means that the resource object will stay in existence until all
+garbage collection mechanisms (owner reference, usage references, TTL) agree it
+should be deleted.
+
+[detailed design of TTL]: https://goo.gl/9AUQ3n
+[detailed design of GC after use (part 1)]: https://goo.gl/ws9t3b
+[detailed design of GC after use (part 2)]: https://goo.gl/KeHPiQ
+
+### Generic TTL-based Garbage Collection
+
+First, we need to clarify what TTL means in this context. TTL represents the
+earliest safe time when a resource can be garbage collected. 
+
+Generic TTL-based garbage collection will allow the user to set a TTL on each of
+Kubernetes resources. The TTL starts when a resource is eligible for garbage
+collection and stops when the resource is no longer eligible; the system will
+only garbage collect the resource after the TTL expires. This is useful when the
+user expects a resource to be needed intermittently and/or want to keep the
+resource around for a while before garbage collection kicks in.
+
+Note that we will not use etcd TTL to support this feature, because etcd TTLs
```

Comment:
It sounds to me that the proposed TTL GC is a generic version of the etcd TTL that works only on events. Is the proposed TTL GC able to replace etcd TTL in future? To make it a step further, I think the proposed TTL GC should have the long-run goal of unifying all GC flavors (OwnerReference, Capacity, and TTL) into one.

Reply:
How we should handle Event TTL is still under discussion. TTL GC won't replace existing Event TTL mechanism in the first (alpha) release.

Reply:
sgtm, thanks!